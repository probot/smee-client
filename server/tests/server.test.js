const nock = require('nock')
const createServer = require('../server')
const request = require('supertest')
const EventSource = require('eventsource')
const cookieSession = require('cookie-session')
const Keygrip = require('keygrip')

describe('server', () => {
  let app, server, events, url, channel

  beforeEach((done) => {
    channel = '/fake-channel'
    app = createServer()
    app.use(cookieSession({
      name: 'session',
      keys: ['key1'],
      secret: process.env.WEBHOOK_SECRET,
      maxAge: 30 * 24 * 60 * 60 * 1000
    }))
    server = app.listen(0, () => {
      url = `http://127.0.0.1:${server.address().port}${channel}`

      // Wait for event source to be ready
      events = new EventSource(url)
      events.addEventListener('ready', () => done())
    })
  })

  afterEach(() => {
    server && server.close()
    events && events.close()
  })

  describe('GET /', () => {
    it('returns the proper HTML when unauthenticated', async () => {
      const res = await request(server).get('/')
      expect(res.status).toBe(200)
    })

    it('returns the proper HTML when authenticated', async () => {
      let cookie = Buffer.from(JSON.stringify({'user': 'test', 'token': 'test'})).toString('base64')
      let kg = Keygrip(['key1'])
      let hash = kg.sign('session=' + cookie)
      const res = await request(server).get('/')
        .set('cookie', ['session=' + cookie + '; ' + 'session.sig=' + hash + ';'])
      expect(res.status).toBe(200)
      expect(res.text).toMatchSnapshot()
    })
  })

  describe('GET /logout', () => {
    it('logout the user', async () => {
      const res = await request(server).get('/logout')
      expect(res.redirect).toBe(true)
    })
  })

  describe('GET /new', () => {
    // setting cookie
    it('redirects to channel if authenticated', async () => {
      let cookie = Buffer.from(JSON.stringify({'user': 'test', 'token': 'test'})).toString('base64')
      let kg = Keygrip(['key1'])
      let hash = kg.sign('session=' + cookie)
      const res = await request(server).get('/new')
        .set('cookie', ['session=' + cookie + '; ' + 'session.sig=' + hash + ';'])
      expect(res.status).toBe(302)
    })

    it('redirects to github', async () => {
      await request(server).get('/new')
        .expect(302)
        .expect('Location', /^https:\/\/github.com\/login\/oauth\/authorize/)
    })
  })

  describe('GET /auth/github/callback', () => {
    it('redirects from github', async () => {
      nock('https://github.com').post('/login/oauth/access_token')
        .reply(200, { access_token: 'test', token_type: 'bearer' })
      await request(server).get('/auth/github/callback')
        .expect(302)
    })
  })

  describe('GET /:channel', () => {
    it('returns the proper HTML when unauthenticated', async () => {
      const res = await request(server).get(channel)
      expect(res.status).toBe(200)
      expect(res.text).toMatchSnapshot()
    })

    it('checks user authentication and redirects', async () => {
      let cookie = Buffer.from(JSON.stringify({'user': 'test', 'token': 'test'})).toString('base64')
      let kg = Keygrip(['key1'])
      let hash = kg.sign('session=' + cookie)
      const res = await request(server).get(channel)
                  .set('cookie', ['session=' + cookie + '; ' + 'session.sig=' + hash + ';'])
      expect(res.status).toBe(307)
      expect(res.redirect).toBe(true)
    })
  })

  describe('GET /:user/:channel', () => {
    it('session not set', async () => {
      const res = await request(server).get('/fake' + channel)
      expect(res.redirect).toBe(true)
    })

    it(' wrong user ', async () => {
      let cookie = Buffer.from(JSON.stringify({'user': 'test', 'token': 'test'})).toString('base64')
      let kg = Keygrip(['key1'])
      let hash = kg.sign('session=' + cookie)
      const res = await request(server).get('/fake' + channel)
        .set('cookie', ['session=' + cookie + '; ' + 'session.sig=' + hash + ';'])
      expect(res.status).toBe(200)
      expect(res.text).toMatchSnapshot()
    })

    it(' same user ', async (done) => {
      let cookie = Buffer.from(JSON.stringify({'user': 'fake', 'token': 'test'})).toString('base64')
      let kg = Keygrip(['key1'])
      let hash = kg.sign('session=' + cookie)
      const newChannel = '/fake-user/fake-channel'
      const newServer = app.listen(0, () => {
        url = `http://127.0.0.1:${server.address().port}${newChannel}`

        // Wait for event source to be ready
        events = new EventSource(url)
        events.addEventListener('ready', () => done())
      })

      const res = await request(newServer).get('/fake' + channel)
        .set('cookie', ['session=' + cookie + '; ' + 'session.sig=' + hash + ';'])
      expect(res.status).toBe(200)
      expect(res.text).toMatchSnapshot()
    })
  })

  describe('events', () => {
    it('emits events', async (done) => {
      const payload = {payload: true}

      await request(server).post(channel)
        .set('X-Foo', 'bar')
        .send(payload)
        .expect(200)

      events.addEventListener('message', (msg) => {
        const data = JSON.parse(msg.data)
        expect(data.body).toEqual(payload)
        expect(data['x-foo']).toEqual('bar')
        done()
      })
    })

    it('POST /:channel/redeliver re-emits a payload', async () => {
      const payload = {payload: true}

      await request(server).post(channel + '/redeliver')
        .send(payload)
        .expect(200)

      events.addEventListener('message', (msg) => {
        const data = JSON.parse(msg.data)
        expect(data).toEqual(payload)
      })
    })

    it('emits events for authenticated', async (done) => {
      const payload = {payload: true}
      const newChannel = '/fake-user/fake-channel'
      const newServer = app.listen(0, () => {
        url = `http://127.0.0.1:${server.address().port}${newChannel}`

        // Wait for event source to be ready
        events = new EventSource(url)
        events.addEventListener('ready', () => done())
      })
      await request(newServer).post('/fake-user' + channel)
        .set('X-Foo', 'bar')
        .send(payload)
        .expect(200)

      events.addEventListener('message', (msg) => {
        const data = JSON.parse(msg.data)
        expect(data.body).toEqual(payload)
        expect(data['x-foo']).toEqual('bar')
        done()
        // test is done if all of this gets calle
      })
    })

    it('POST /:user/:channel/redeliver re-emits a payload', async (done) => {
      const payload = {payload: true}
      const newChannel = '/fake-user/fake-channel'
      const newServer = app.listen(0, () => {
        url = `http://127.0.0.1:${server.address().port}${newChannel}`

        // Wait for event source to be ready
        events = new EventSource(url)
        events.addEventListener('ready', () => done())
      })
      await request(newServer).post(newChannel + '/redeliver')
        .send(payload)
        .expect(200)

      events.addEventListener('message', (msg) => {
        const data = JSON.parse(msg.data)
        expect(data).toEqual(payload)
        done()
      })
    })
  })
})
