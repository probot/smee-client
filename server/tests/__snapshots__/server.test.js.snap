// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server GET / returns the proper HTML when authenticated 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
  <meta charset=\\"UTF-8\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
  <title>smee.io</title>
  <link rel=\\"shortcut icon\\" href=\\"/public/favicon.png\\">
  <link rel=\\"stylesheet\\" href=\\"/public/main.min.css\\">
</head>
<body>
  <header class=\\"text-white text-center p-responsive bg-blue d-flex flex-column flex-items-center flex-justify-center\\" style=\\"height: 80vh\\">
    <a href=\\"/logout\\" class=\\"btn btn-outline btn-outline-red position\\">Logout</a>
    <div class=\\"header__anim\\">
      <div class=\\"header__anim__payload\\"></div>
      <div class=\\"header__anim__line\\"></div>
      <div class=\\"header__anim__circle header__anim__center\\"></div>
      <div class=\\"header__anim__circle header__anim__dashed-circle\\"></div>
    </div>
    <h1 class=\\"f00-light\\">smee.io</h1>
    <h2 class=\\"blue-700\\">Webhook payload delivery service</h2>
    <p class=\\"lead text-white\\" style=\\"opacity: 0.8\\">Receives payloads then sends them to your locally running application.</p>
    <a href=\\"/new\\" class=\\"btn btn-outline btn-outline-blue\\">Start a new channel</a>
  </header>

  <main class=\\"container-lg py-6 mt-6 p-responsive\\">
    <p class=\\"lead text-center col-12 col-md-8 mx-auto\\">If your application needs to respond to webhooks, you'll need some way to expose <strong>localhost</strong> to the internet. <strong>smee.io</strong> is a small service that uses <a href=\\"https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">Server-Sent Events</a> to proxy payloads from the webhook source, then transmit them to your locally running application.</p>

    <div class=\\"main__anim my-6\\">
      <div class=\\"main__anim__circle main__anim__left\\">
        <span>Webhook Emitter</span>
      </div>
      <div class=\\"main__anim__circle main__anim__right\\">
        <span>localhost</span>
      </div>
      <div class=\\"main__anim__line\\"></div>
      <div class=\\"main__anim__payload\\"></div>
      <div class=\\"main__anim__circle main__anim__center\\">
        <span></span>
      </div>
    </div>

    <p class=\\"lead text-center col-12 col-md-8 mx-auto\\">Tell your webhook source to send payloads to your <strong>smee.io</strong> channel, then either use the <a href=\\"https://github.com/probot/smee/tree/master/client\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">smee client</a> or, if you're using <a href=\\"https://probot.github.io\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">Probot</a> to build a GitHub App, just <a href=\\"https://probot.github.io/docs/development/#configure-a-github-app\\">set the environment variable</a>.</p>
  </main>

  <footer class=\\"border-top text-center py-4\\">
    Made with ðŸ¤– by the <a href=\\"https://github.com/probot\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">Probot team</a>.
  </footer>

  <script async src=\\"https://www.googletagmanager.com/gtag/js?id=UA-102577034-3\\"></script>
  <script>
    window.dataLayer = window.dataLayer || [];
    function gtag(){dataLayer.push(arguments);}
    gtag('js', new Date());
    gtag('config', 'UA-102577034-3');
  </script>
</body>
</html>
"
`;

exports[`server GET / returns the proper HTML when unauthenticated 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
  <meta charset=\\"UTF-8\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
  <title>smee.io</title>
  <link rel=\\"shortcut icon\\" href=\\"/public/favicon.png\\">
  <link rel=\\"stylesheet\\" href=\\"/public/main.min.css\\">
</head>
<body>
  <header class=\\"text-white text-center p-responsive bg-blue d-flex flex-column flex-items-center flex-justify-center\\" style=\\"height: 80vh\\">
    <div class=\\"header__anim\\">
      <div class=\\"header__anim__payload\\"></div>
      <div class=\\"header__anim__line\\"></div>
      <div class=\\"header__anim__circle header__anim__center\\"></div>
      <div class=\\"header__anim__circle header__anim__dashed-circle\\"></div>
    </div>
    <h1 class=\\"f00-light\\">smee.io</h1>
    <h2 class=\\"blue-700\\">Webhook payload delivery service</h2>
    <p class=\\"lead text-white\\" style=\\"opacity: 0.8\\">Receives payloads then sends them to your locally running application.</p>
    <a href=\\"/new\\" class=\\"btn btn-outline btn-outline-green\\">Log in to start a new private channel</a>
  </header>

  <main class=\\"container-lg py-6 mt-6 p-responsive\\">
    <p class=\\"lead text-center col-12 col-md-8 mx-auto\\">If your application needs to respond to webhooks, you'll need some way to expose <strong>localhost</strong> to the internet. <strong>smee.io</strong> is a small service that uses <a href=\\"https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">Server-Sent Events</a> to proxy payloads from the webhook source, then transmit them to your locally running application.</p>

    <div class=\\"main__anim my-6\\">
      <div class=\\"main__anim__circle main__anim__left\\">
        <span>Webhook Emitter</span>
      </div>
      <div class=\\"main__anim__circle main__anim__right\\">
        <span>localhost</span>
      </div>
      <div class=\\"main__anim__line\\"></div>
      <div class=\\"main__anim__payload\\"></div>
      <div class=\\"main__anim__circle main__anim__center\\">
        <span></span>
      </div>
    </div>

    <p class=\\"lead text-center col-12 col-md-8 mx-auto\\">Tell your webhook source to send payloads to your <strong>smee.io</strong> channel, then either use the <a href=\\"https://github.com/probot/smee/tree/master/client\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">smee client</a> or, if you're using <a href=\\"https://probot.github.io\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">Probot</a> to build a GitHub App, just <a href=\\"https://probot.github.io/docs/development/#configure-a-github-app\\">set the environment variable</a>.</p>
  </main>

  <footer class=\\"border-top text-center py-4\\">
    Made with ðŸ¤– by the <a href=\\"https://github.com/probot\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">Probot team</a>.
  </footer>

  <script async src=\\"https://www.googletagmanager.com/gtag/js?id=UA-102577034-3\\"></script>
  <script>
    window.dataLayer = window.dataLayer || [];
    function gtag(){dataLayer.push(arguments);}
    gtag('js', new Date());
    gtag('config', 'UA-102577034-3');
  </script>
</body>
</html>
"
`;

exports[`server GET /:user/:channel  wrong user  1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
  <meta charset=\\"UTF-8\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
  <title>smee.io</title>
  <link rel=\\"shortcut icon\\" href=\\"/public/favicon.png\\">
  <link rel=\\"stylesheet\\" href=\\"/public/main.min.css\\">
</head>
<body>
  <div class=\\"svg-container\\">
    <div class=\\"col-xs-12 col-md-12\\">
      <svg viewBox=\\"0 0 1581.000000 760.000000\\" style=\\"width: 100%;padding-bottom: 92%;height:0px;overflow: visible;\\">
          <g transform=\\"translate(0.000000,760.000000) scale(0.100000,-0.100000)\\"
          fill=\\"grey\\" stroke=\\"none\\">
          <path d=\\"M8123 4389 c-21 -11 -36 -39 -47 -91 -5 -24 -15 -30 -74 -48 -101
          -30 -208 -104 -407 -280 -38 -34 -87 -71 -107 -81 -39 -20 -53 -47 -46 -90 4
          -28 31 -36 64 -18 21 11 26 10 39 -7 13 -16 16 -17 25 -4 5 8 10 23 10 33 0
          43 198 207 249 207 19 0 8 -38 -42 -145 l-42 -90 0 -417 c0 -411 0 -418 -20
          -419 -19 0 -19 -1 1 -9 11 -4 26 -22 33 -39 9 -21 22 -32 44 -37 18 -3 39 -12
          48 -20 9 -8 50 -18 93 -24 43 -5 82 -12 87 -15 5 -4 9 -1 9 5 0 6 -25 34 -55
          62 -30 27 -55 54 -55 58 0 5 10 11 23 13 20 4 19 4 -4 6 -31 1 -36 14 -45 111
          -8 82 8 209 45 375 17 72 33 159 37 195 8 77 51 204 106 315 57 114 69 125
          137 125 31 0 73 5 93 11 46 12 52 31 46 138 -5 91 -27 130 -90 162 -47 24
          -125 33 -155 18z m167 -194 c0 -28 -17 -42 -33 -26 -7 7 -7 18 1 36 14 30 32
          24 32 -10z\\"/>
          <path d=\\"M7920 2965 c-69 -7 -125 -13 -125 -14 0 -6 -51 -10 -155 -12 -121 -1
          -186 -14 -434 -84 -186 -52 -270 -98 -383 -209 -35 -33 -63 -66 -63 -73 0 -17
          16 -17 23 0 3 8 31 31 64 52 32 21 81 56 108 76 28 21 86 52 130 69 44 17 85
          36 90 40 13 11 178 59 220 65 16 2 86 14 155 27 107 19 157 22 340 22 273 -1
          432 -23 500 -68 19 -13 67 -36 105 -50 104 -39 121 -49 235 -141 58 -47 113
          -85 123 -85 36 0 14 29 -60 80 -43 30 -99 72 -124 94 -99 88 -113 96 -240 136
          -70 22 -135 40 -144 40 -9 0 -31 10 -48 23 -39 27 -130 31 -317 12z\\"/>
          <path d=\\"M15700 2964 c-108 -40 -205 -82 -422 -186 -259 -123 -376 -171 -603
          -247 -586 -197 -1170 -282 -1720 -251 -516 28 -699 66 -1720 350 -175 49 -507
          130 -650 159 -576 116 -1004 63 -1649 -204 -68 -28 -111 -39 -184 -46 -67 -6
          -101 -14 -115 -26 -10 -10 -28 -17 -38 -15 -10 2 -19 -1 -19 -7 0 -15 -62 -45
          -145 -70 -110 -34 -417 -84 -525 -84 -90 -1 -99 1 -167 36 -39 20 -91 39 -115
          42 -37 5 -34 3 20 -14 76 -22 101 -41 56 -41 -85 0 -312 45 -423 85 -24 8 -63
          15 -86 15 -30 0 -136 34 -342 109 -695 253 -901 301 -1233 288 -242 -10 -352
          -43 -673 -196 -397 -191 -561 -229 -917 -217 -410 15 -562 41 -1310 230 -524
          132 -864 197 -1200 228 -220 20 -972 17 -1210 -5 -247 -23 -288 -32 -301 -63
          -5 -16 -9 -573 -9 -1430 l0 -1404 7895 0 7895 0 0 1495 c0 1190 -3 1495 -12
          1494 -7 -1 -42 -12 -78 -25z\\" class=\\"upfill\\"/>

          </g>
      </svg>
    </div>
  </div>

  <div class=\\"container-md\\">
    <div class=\\"row\\">
      <div class=\\"col-xs-12 col-md-12 text-center\\">
        <div style=\\"color:#83668A;position:fixed;top:10%;\\">
          <h1>This page was lost.</h1>
          <h1>Looks like you don't have authorization to this URL.</h1>
        </div>
      </div>
    </div>
    <div class=\\"row\\" style=\\"position:fixed;bottom:20%;font-size:22px;color:white\\">
      <div class=\\"col-xs-12 col-md-12\\">
        <a href=\\"/\\" class=\\"btn btn-outline btn-outline-green\\">Home Page</a>
        <a href=\\"/logout\\" class=\\"btn btn-outline btn-outline-red\\">Logout</a>
      </div>
    </div>
    <div class=\\"row\\">
      <div class=\\"col-xs-12 col-md-12\\">
        <footer class=\\"text-center py-6\\" style=\\"position:fixed;bottom:5%;font-size:22px;color:white\\">
          Made with ðŸ¤– by the <a href=\\"https://github.com/probot\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\" style=\\"color:	#BB4889\\">Probot team</a>.
        </footer>
      </div>
    </div>
  </div>

</body>
</html>
"
`;

exports[`server events emits events 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
  <meta charset=\\"UTF-8\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
  <title>Webhooks</title>
  <link rel=\\"shortcut icon\\" href=\\"/public/favicon.png\\">
  <link rel=\\"stylesheet\\" href=\\"/public/main.min.css\\">
</head>
<body>
  <noscript>
    <div class=\\"no-script\\">
      <h1>You must enable JavaScript to use try out todo.</h1>
      <a href=\\"https://github.com/probot/webhooks/issues\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">Need help?</a>
    </div>
  </noscript>
  <div class=\\"mount\\"></div>
  <script src=\\"/public/main.min.js\\"></script>
  <script async src=\\"https://www.googletagmanager.com/gtag/js?id=UA-102577034-3\\"></script>
  <script>
    window.dataLayer = window.dataLayer || [];
    function gtag(){dataLayer.push(arguments);}
    gtag('js', new Date());
    gtag('config', 'UA-102577034-3');
  </script>
</body>
</html>
"
`;
